src/lib/server/write_buffer.cpp

```diff
@@ -0,0 +1,92 @@
+#include "write_buffer.hpp"
+
+#include <pthread.h>
+
+namespace opossum {
+
+template <typename SocketType>
+size_t WriteBuffer<SocketType>::size() const {
+  const auto current_size = std::distance(&*_start_position, &*_current_position);
+  return current_size < 0 ? current_size + BUFFER_SIZE : current_size;
+}
+
+template <typename SocketType>
+size_t WriteBuffer<SocketType>::maximum_capacity() const {
+  return BUFFER_SIZE - 1;
+}
+
+template <typename SocketType>
+bool WriteBuffer<SocketType>::full() const {
+  return size() == maximum_capacity();
+}
+
+template <typename SocketType>
+void WriteBuffer<SocketType>::put_string(const std::string& value, const HasNullTerminator has_null_terminator) {
+  auto position_in_string = 0u;
+
+  // Use available space first
+  if (!full()) {
+    position_in_string = static_cast<uint32_t>(std::min(maximum_capacity() - size(), value.size()));
+    std::copy_n(value.cbegin(), position_in_string, _current_position);
+    std::advance(_current_position, position_in_string);
+  }
+
+  // Write to network device until string is complete. Ignore last character since it is \0
+  while (position_in_string < value.size()) {
+    const auto bytes_to_transfer = std::min(maximum_capacity(), value.size() - position_in_string);
+    _flush_if_necessary(bytes_to_transfer);
+    std::copy_n(value.cbegin() + position_in_string, bytes_to_transfer, _current_position);
+    std::advance(_current_position, bytes_to_transfer);
+    position_in_string += bytes_to_transfer;
+  }
+
+  // Add string terminator if necessary
+  if (has_null_terminator == HasNullTerminator::Yes) {
+    _flush_if_necessary(sizeof(char));
+    *_current_position = '\0';
+    _current_position++;
+  }
+}
+
+template <typename SocketType>
+void WriteBuffer<SocketType>::flush(const size_t bytes_required) {
+  const auto bytes_to_send = bytes_required ? bytes_required : size();
+  size_t bytes_sent;
+
+  boost::system::error_code error_code;
+  if (std::distance(&*_start_position, &*_current_position) < 0) {
+    // Data not continuously stored in buffer
+    bytes_sent =
+        boost::asio::write(*_socket,
+                           std::array<boost::asio::mutable_buffer, 2>{
+                               boost::asio::buffer(&*_start_position, std::distance(&*_start_position, _data.end())),
+                               boost::asio::buffer(_data.begin(), std::distance(_data.begin(), &*_current_position))},
+                           boost::asio::transfer_at_least(bytes_to_send), error_code);
```

Comment:
I don‘t get this yet. Bytes_to_send can be larger than size - how can I flush data that I don‘t have yet?

Reply:
You are right, that shouldn't be possible. Adding an Assert...